examples/fastapi-embeddings-rag/app/main.py

```diff
@@ -0,0 +1,379 @@
+# Runhouse [`env`](https://www.run.house/docs/tutorials/api-envs) so that they can be properly installed
+# on the cluster. Also note that they are imported inside class methods. Those packages do not need
+# to be installed locally.
+def load_embedder():
+    """Launch an A10G and send the embedding service to it."""
+    cluster = rh.cluster(
+        CLUSTER_NAME, instance_type=INSTANCE_TYPE, provider=CLOUD_PROVIDER
+    ).up_if_not()
+
+    env = rh.env(
+        name="embedder_env",
+        reqs=[
+            "langchain",
+            "langchain-community",
+            "langchain_text_splitters",
+            "langchainhub",
+            "bs4",
+            "sentence_transformers",
+            "torch",
+        ],
+        compute={"GPU": 1},
+    )
+
+    module_name = "url_embedder"
+    remote_url_embedder = cluster.get(module_name, default=None, remote=True)
+    if DEBUG or remote_url_embedder is None:
+        RemoteEmbedder = rh.module(URLEmbedder).to(
+            system=cluster, env=env, name="URLEmbedder"
+        )
+        remote_url_embedder = RemoteEmbedder(
+            model_name_or_path="BAAI/bge-large-en-v1.5", device="cuda", name=module_name
+        )
+    return remote_url_embedder
+
+
+# ### Initialize LanceDB Vector Database
+# We'll be using open source [LanceDB](https://lancedb.com/) to create an embedded database to store
+# the URL embeddings and perform vector search for the retrieval phase. You could alternatively try
+# Chroma, Pinecone, Weaviate, or even MongoDB.
+def load_table():
+    # Initialize LanceDB database directly on the FastAPI app's machine
+    db = lancedb.connect("/tmp/db")
+    return db.create_table("rag-table", schema=Item.to_arrow_schema(), exist_ok=True)
+
+
+# ### Load RAG LLM Inference Service with Runhouse
+# Deploy an open LLM, Llama 3 in this case, to a GPU on the cloud provider of your choice.
+# We will use vLLM to serve the model due to it's high performance and throughput but there
+# are many other options such as HuggingFace Transforms and TGI.
+#
+# Here we leverage the same A10G cluster we used for the embedding service, but you could also spin
+# up a new remote machine specifically for the LLM service. Alternatively, use a proprietary model
+# like ChatGPT or Claude.
+def load_llm():
+    """Use the existing A10G cluster to run an LLM inference service"""
+    # Specifying the same name will reuse our embedding service cluster
+    cluster = rh.cluster(
+        CLUSTER_NAME, instance_type=INSTANCE_TYPE, provider=CLOUD_PROVIDER
+    ).up_if_not()
+
+    env = rh.env(
+        reqs=["torch", "vllm"],
+        secrets=["huggingface"],  # Needed to download Llama 3 from HuggingFace
+        name="llama3_inference_env",
+    )
+
+    module_name = "llama_model"
+    # First check for an instance of the LlamaModel stored on the cluster
+    remote_llm = cluster.get(module_name, default=None, remote=True)
```

Comment:
is this block here to remove confusion behind `get_or_to`?

Reply:
yeah and I think `.get_or_to` isn't necessarily working as intended for modules? @dongreenberg made this point last week